### PR TITLE
[dbnode] Only allocate index block compactors before use

### PR DESCRIPTION
**Pull request by robskillington**
This avoids the excessive memory usage when a lot of index blocks are created and compactors with a two roaring bitmaps with up front allocated buckets take up considerable memory usage.

### CODE REVIEW

File: src/dbnode/storage/index/block.go [hunk 1]
```diff
@@ -799,7 +813,8 @@ func (b *block) Query(
 	}
 
 	size := results.Size()
-	batch := b.docsPool.Get()
+	docsPool := b.opts.DocumentArrayPool()
```
Comment (richardartoul) [inline src/dbnode/storage/index/block.go]: Super nit: `var` block would be nice here.

File: src/dbnode/storage/index/block.go [hunk 2]
```diff
@@ -110,10 +110,10 @@ type block struct {
 	blockStart    time.Time
 	blockEnd      time.Time
 	blockSize     time.Duration
+	blockOpts     BlockOptions
 	opts          Options
 	iopts         instrument.Options
 	nsMD          namespace.Metadata
-	docsPool      doc.DocumentArrayPool
```
Comment (prateek) [inline src/dbnode/storage/index/block.go]: super nit: mind grouping the lazily allocated objects are grouped together in a struct and there's a comment about using the ensure method before using

Comment (robskillington) [reply src/dbnode/storage/index/block.go]: Good call.

File: src/dbnode/storage/index/block.go (hunk 1 above)
Comment (robskillington) [reply src/dbnode/storage/index/block.go]: Sure yeah.

**Pull request merged**